app-modules/meeting-center/src/GoogleCalendarManager.php

```diff
@@ -0,0 +1,205 @@
+        } else {
+            $event = $this->createEvent($event);
+        }
+
+        return $event;
+    }
+
+    public function deleteEvent(CalendarEvent $event): void
+    {
+        $service = (new GoogleCalendar(static::client($event->calendar)));
+        $service->events->delete($event->calendar->provider_id, $event->provider_id);
+    }
+
+    public function syncEvents(Calendar $calendar): void
+    {
+        $events = collect($this->getEvents($calendar));
+
+        $events
+            ->each(
+                function (Event $event) use ($calendar) {
+                    $userEvent = $calendar->events()->where('provider_id', $event->id)->first();
+
+                    if ($userEvent) {
+                        $userEvent
+                            ->updateQuietly([
+                                'title' => $event->summary,
+                                'description' => $event->description,
+                                'starts_at' => $event->start->dateTime,
+                                'ends_at' => $event->end->dateTime,
+                            ]);
+                    } else {
+                        $calendar
+                            ->events()
+                            ->createQuietly([
+                                'provider_id' => $event->id,
+                                'title' => $event->summary,
+                                'description' => $event->description,
+                                'starts_at' => $event->start->dateTime,
+                                'ends_at' => $event->end->dateTime,
+                            ]);
+                    }
+                }
+            );
+
+        $calendar->events()
+            ->whereNull('provider_id')
+            ->each(fn ($event) => $this->createEvent($event)->saveQuietly());
+
+        $calendar->events()->whereNotIn('provider_id', $events->pluck('id'))->delete();
+    }
+
+    public static function client(?Calendar $calendar = null): Client
+    {
+        if ($calendar?->oauth_token) {
+            $client = new Client([
+                'client_id' => config('services.google_calendar.client_id'),
+                'client_secret' => config('services.google_calendar.client_secret'),
+                'scopes' => [
+                    GoogleCalendar::CALENDAR,
+                    GoogleCalendar::CALENDAR_EVENTS,
+                ],
+            ]);
+
+            if ($calendar->oauth_token_expiress_at < now()) {
+                $token = $client->fetchAccessTokenWithRefreshToken($calendar->oauth_refresh_token);
+                $calendar->oauth_token = $token['access_token'];
+                $calendar->oauth_token_expires_at = Carbon::parse($token['created'] + $token['expires_in']);
+                $calendar->save();
+            } else {
+                $client->setAccessToken($calendar->oauth_token);
+            }
```

Comment:
Should `$client->setAccessToken($calendar->oauth_token);` be outside of the `else`? or does `fetchAccessTokenWithRefreshToken` handle setting the new on in the client for us?

Reply:
Yeah it's a little weird, but `fetchAccessTokenWithRefreshToken` automatically updates the object, which is why I just update our database copy.

```php
public function fetchAccessTokenWithRefreshToken($refreshToken = null)
{
    if (null === $refreshToken) {
        if (!isset($this->token['refresh_token'])) {
            throw new LogicException(
                'refresh token must be passed in or set as part of setAccessToken'
            );
        }
        $refreshToken = $this->token['refresh_token'];
    }
    $this->getLogger()->info('OAuth2 access token refresh');
    $auth = $this->getOAuth2Service();
    $auth->setRefreshToken($refreshToken);

    $httpHandler = HttpHandlerFactory::build($this->getHttpClient());
    $creds = $auth->fetchAuthToken($httpHandler);
    if ($creds && isset($creds['access_token'])) {
        $creds['created'] = time();
        if (!isset($creds['refresh_token'])) {
            $creds['refresh_token'] = $refreshToken;
        }
        $this->setAccessToken($creds);
    }

    return $creds;
}
```